File: tests/xcalablemp/global-view/gmove/C/gmove_ata_3a1t_b2.c
#include <stdio.h>
#include <math.h>

extern int chk_int(int ierr);

int n=8;
double a[n][n][n],b[n][n][n];
#pragma xmp nodes p(2)
#pragma xmp template tx(0:n-1)
#pragma xmp distribute tx(block) onto p
#pragma xmp align a[i][*][*] with tx(i)
#pragma xmp align b[*][*][i] with tx(i)

int main(){

  int i,j,k,ierr;
  double err;

#pragma xmp loop (i) on tx(i)
  for(i=0;i<n;i++){
    for(j=0;j<n;j++){
      for(k=0;k<n;k++){
        a[i][j][k]=i+j+k+2;
      }
    }
  }

  for(i=0;i<n;i++){
    for(j=0;j<n;j++){
#pragma xmp loop (k) on tx(k)
      for(k=0;k<n;k++){
        b[i][j][k]=0;
      }
    }
  }

#pragma xmp gmove
  b[:][:][:]=a[:][:][:];

  err=0.0;
#pragma xmp loop (i) on tx(i)
  for(i=0;i<n;i++){
    for(j=0;j<n;j++){
#pragma xmp loop (k) on tx(k)
      for(k=0;k<n;k++){
/*        printf("i=%d,j=%d,k=%d,b=%f,a=%f\n",i,j,k,b[i][j][k],a[i][j][k],myrank);i*/
        err=err+fabs(b[i][j][k]-a[i][j][k]);
      }
    }
  }

#pragma xmp reduction (MAX:err)
  ierr=err;
  chk_int(ierr);

}

File: tests/xcalablemp/global-view/gmove/Fortran/gmove_ata_3a1t_b2.f90
program tgmove
integer :: i0,i1,i2
integer,parameter :: n=4
integer a(n,n,n),b(n,n,n)
integer xmp_node_num
!$xmp nodes p(2)
!$xmp template tx(n)
!$xmp distribute tx(block) onto p
!$xmp align a(i,*,*) with tx(i)
!$xmp align b(*,i,*) with tx(i)

!$xmp loop (i0) on tx(i0)
do i2=1,n
  do i1=1,n
    do i0=1,n
      a(i0,i1,i2)=i0+i1+i2
    end do
  end do
end do

do i2=1,n
!$xmp loop (i1) on tx(i1)
  do i1=1,n
    do i0=1,n
      b(i0,i1,i2)=0
    end do
  end do
end do

!$xmp gmove
b(:,:,:)=a(:,:,:)

ierr=0
do i2=1,n
!$xmp loop (i1) on tx(i1)
  do i1=1,n
!$xmp loop (i0) on tx(i0)
    do i0=1,n
      ierr=ierr+abs(b(i0,i1,i2)-a(i0,i1,i2))
    end do
  end do
end do

!$xmp reduction (max:ierr)
irank=xmp_node_num()
if (irank==1) then
  print *, 'max error=',ierr
endif
!call exit(ierr)

stop
end program tgmove
